pkg/graphite/convert/whisperconverter/whisper.go

```diff
@@ -40,67 +41,85 @@ type Archive interface {
+	archives := w.GetArchives()
+	if len(archives) == 0 {
 		return nil, fmt.Errorf("whisper file contains no archives for metric: %q", name)
 	}
 
-	totalPoints := 0
-	for _, a := range w.GetArchives() {
-		totalPoints += int(a.Points)
-	}
-
-	// Preallocate space for all allPoints in one slice.
-	allPoints := make([]pointWithPrecision, totalPoints)
-	pIdx := 0
 	// Dump one precision level at a time and write into the output slice.
-	for i, a := range w.GetArchives() {
-		archivePoints, err := w.DumpArchive(i)
+	// Its important to remember that the archive with index 0 (first archive)
+	// has the raw data and the highest precision https://graphite.readthedocs.io/en/latest/whisper.html#archives-retention-and-precision
+	var keptPoints []whisper.Point
+
+	// We want to track the max timestamp of the archives because we know
+	// it virtually represents now() and we wont have newer points.
+	// Then the min timestamp of the archive would be maxTs - each archive
+	// retention.
+	var maxTs uint32
+	lastMinTs := uint32(math.MaxUint32)
+
+	for i, a := range archives {
+		points, err := w.DumpArchive(i)
 		if err != nil {
 			return nil, fmt.Errorf("failed to dump archive %d from whisper metric %s", i, name)
 		}
-		for j, p := range archivePoints {
-			allPoints[pIdx+j] = pointWithPrecision{p, a.SecondsPerPoint}
+
+		if len(points) == 0 {
+			continue
 		}
-		pIdx += len(archivePoints)
-	}
 
-	// Points must be in time order.
-	sort.Slice(allPoints, func(i, j int) bool {
-		return allPoints[i].Timestamp < allPoints[j].Timestamp
-	})
+		// All archives share the same maxArchiveTs, so only calculate it once.
+		if maxTs == 0 {
+			for _, p := range points {
+				if p.Timestamp > maxTs {
+					maxTs = p.Timestamp
+				}
+			}
+		}
 
-	trimmedPoints := []whisper.Point{}
-	for i := 0; i < len(allPoints); i++ {
-		// Remove all points of time = 0.
-		if allPoints[i].Timestamp == 0 {
-			continue
+		var minArchiveTs uint32
+		if maxTs < a.Retention() { // very big retention, invalid.
+			minArchiveTs = 0
+		} else {
+			minArchiveTs = maxTs - a.Retention()
 		}
-		// There might be duplicate timestamps in different archives. Take the
-		// higher-precision archive value since it's unaggregated.
-		if i > 0 && allPoints[i].Timestamp == allPoints[i-1].Timestamp {
-			if allPoints[i].secondsPerPoint == allPoints[i-1].secondsPerPoint {
-				return nil, fmt.Errorf("duplicate timestamp at same precision in archive %s: %d", name, allPoints[i].Timestamp)
+
+		// Sort this archive.
+		sort.Slice(points, func(i, j int) bool {
+			return points[i].Timestamp < points[j].Timestamp
+		})
+
+		startIdx := -1
+		endIdx := len(points) - 1
+		for j, p := range points {
+			if p.Timestamp == 0 {
+				continue
 			}
-			if allPoints[i].secondsPerPoint < allPoints[i-1].secondsPerPoint {
-				trimmedPoints[len(trimmedPoints)-1] = allPoints[i].Point
+			// Don't include any points in this archive that are past the retention
+			// period.
+			if p.Timestamp < minArchiveTs {
```

Comment:
Nit: I think it should be

```suggestion
			if p.Timestamp <= minArchiveTs {
```

Point with Timestamp equal to `minArchiveTs` is already outside of retention.